Use client.is_complete() to determine if a code cell is complete

dsblank opened the pull request:
or needs more code. Also returns indent, but doesn't use it yet.

Closes #11

dsblank commented:
It would be great if someone could check this out. Some points/questions:

* Should the system fallback to inputsplitter?
* what exception should be raised for unknown response?
* the ipython kernel appears to work with these changes, but I didn't explore why
* used https://github.com/jupyter/qtconsole/pull/20 as a reference, but not sure if this handles the blocking kernel issue

dsblank commented:
I should also mention that Jupyter languages that aren't somewhat similar to Python (eg, uses parens) can't use the jupyter-console without some kind of fix, so this is somewhat important.

takluyver commented:
I'm intending to look over this, but I'm a bit preoccupied with a trip to
Geneva at the moment ;-)

takluyver commented on jupyter_console/interactiveshell.py:
```diff
@@ -220,6 +224,43 @@ def handle_execute_reply(self, msg_id, timeout=None):
+            try:
+                msg = self.client.shell_channel.get_msg(block=False, timeout=timeout)
+            except Empty:
+                continue
```
If the kernel doesn't reply, e.g. because it hasn't implemented `is_complete` handling, this will hang forever. After a reasonable (and possibly configurable) timeout, it needs to proceed as if it had got 'unknown' back.

dsblank replied:
I found that sometimes a kernel wasn't quite ready to send a response (or maybe it was delayed). So, I added a method to try a few times before giving up.

takluyver commented on jupyter_console/interactiveshell.py:
```diff
@@ -220,6 +224,43 @@ def handle_execute_reply(self, msg_id, timeout=None):
+        elif status == "incomplete":
+            return True, indent
+        elif status == "invalid":
+            raise SyntaxError()
```
Since we're only doing this so it can be caught and turned into `more = False`, maybe it should just return `False, ''` directly?

dsblank replied:
Restructured a bit.

takluyver commented on jupyter_console/interactiveshell.py:
```diff
@@ -220,6 +224,43 @@ def handle_execute_reply(self, msg_id, timeout=None):
+        elif status == "unknown":
+            return True, indent
+        else:
+            raise Exception("")
```
1. Meaningful errors please!
2. I don't think anything would catch this error, so this means that a malformed message from the kernel would make the frontend crash (I'm sure there are other ways to malform a message that will have that effect, but that doesn't make it a good idea). A warning is probably more appropriate than an error here.

dsblank replied:
Fixed.

takluyver commented on jupyter_console/interactiveshell.py:
```diff
@@ -597,3 +637,21 @@ def init_history(self):
+
+    def handle_source(self, line):
+        """
+        more, indent
```
The docstring could perhaps be a tiny bit more descriptive ;-)

dsblank replied:
Fixed.

takluyver commented on jupyter_console/interactiveshell.py:
```diff
@@ -597,3 +637,21 @@ def init_history(self):
         self.history_manager = ZMQHistoryManager(client=self.client)
         self.configurables.append(self.history_manager)
 
+    def get_source_and_reset(self):
```
Since this is two simple lines of code and it's called in two places, I think it would be clearer to inline it.

dsblank replied:
I had left this so that it could also fallback to the input-splitter. In fact, I added a flag "use_kernel_is_complete" to allow the frontend to use the built-in (old) version, and not ask the kernel. This could also be a workaround if a kernel hasn't implemented this yet, or if it is incorrect.

takluyver replied:
This is back to two lines again, so I'd still like it to be inlined. :-)

dsblank replied:
I think this is a fine DIY abstraction. Pre-mature optimization? Having written this kind of code before, I can think of reasons one might want to add other code in here, or overload it.

takluyver replied:
I still think it would be more readable inline (the performance difference is trivial, I'm not trying to optimise it), but I'm happy enough to give you the author's privilege to determine matters of style.

takluyver commented on jupyter_console/interactiveshell.py:
```diff
@@ -597,3 +637,21 @@ def init_history(self):
+        """
+        ## Ask client if line is complete; get indent for next line:
+        if self._source:
+            self._source += "\n"
```
I think it would be neater to store the input as a list, and then call `self.client.is_complete('\n'.join(self._source_lines_buffered))`

dsblank replied:
Done.

takluyver commented on jupyter_console/interactiveshell.py:
```diff
@@ -220,6 +224,43 @@ def handle_execute_reply(self, msg_id, timeout=None):
+                    frontend should also handle the kernel not replying promptly. It
+                    may default to sending the code for execution, or it may implement
+                    simple fallback heuristics for whether to execute the code
+                    (e.g. execute after a blank line).
```
This looks like a copy-paste from the docs. Since these statuses are handled within this function, and not exposed outside it, I don't think they need to be described in the docstring.

dsblank replied:
Yep.

takluyver commented:
This mostly looks OK to me; I've made a few comments inline.

dsblank commented:
Great; thanks! I didn't want to go too far down a path if I was headed in the wrong direction. I'll address these issues.

dsblank commented:
This should be ready for final review. Thanks!

dsblank commented:
Ping @takluyver ... I think that this is ready.

takluyver commented on jupyter_console/interactiveshell.py:
```diff
@@ -109,12 +109,23 @@ class ZMQTerminalInteractiveShell(TerminalInteractiveShell):
+        handling. If False, then the frontend will use its
+        own is_complete handler.
+        """
+    )
```
I don't think we should have a config option for this: the old way of doing it in the frontend assumes that the code being entered is in Python, and most cases where the kernel doesn't support `is_complete` are non Python. So it's the wrong choice to offer the user.

For now, it's probably sensible to have some way to disable the `is_complete` requests, because the delay for kernels that don't implement it might be annoying. But I would do something much simpler than `input_splitter` - either always execute when a line is entered, or prompt for another line until the last line is blank (i.e. you press enter twice to execute).

Eventually it would be good for the frontend to be able to tell when the kernel doesn't support `is_complete` and activate such fallback behaviour automatically, but that needn't be done now.

takluyver replied:
The name of the config flag seems a bit opaque - it's probably not something anyone will search for if they're seeing a slowdown - but I can't think of a better one at the moment.

takluyver commented on jupyter_console/interactiveshell.py:
```diff
@@ -220,6 +231,47 @@ def handle_execute_reply(self, msg_id, timeout=None):
+        msg = None
+        count = 0
+        max_tries = 5
+        while self.client.is_alive() and count < max_tries:
```
I don't quite like this loop. For one thing, it means that the actual timeout is 5 times the value you pass in for `timeout`. But also, since we're using a blocking client, there shouldn't be any other messages coming back on the shell socket before your reply. If there are, something has gone wrong, and as this will consume and discard messages, it may obscure information useful for finding out what actually happened.

I think we should call `get_msg` just once, without the loop, and have a loud warning if it's something other than our reply.

takluyver commented on jupyter_console/interactiveshell.py:
```diff
@@ -220,6 +231,47 @@ def handle_execute_reply(self, msg_id, timeout=None):
+            count += 1
+        ## Handle response:
+        if count == max_tries:
+            warn('The kernel did not respond to an is_complete_request.')
```
Is this user visible? I suspect there are enough kernels not yet implementing `is_complete` that it will get annoying to see this every time.

dsblank commented:
Ok @takluyver I have addressed all of the issues, and feel that this is much better. 

I had misunderstood the block on the get_msg(), and that was the source of the required loop. Now it works as it should. The config flag now is used to flip between built-in, simple end-of-cell criteria (newline) as you suggested (In fact, had this been the default, I probably wouldn't have written this and probably just went with that). 

Additionally, you were right that users would have seen the message repeatedly. As such, the right thing is to automatically turn off attempting to use the kernel for is_complete, and turn on the automatic method.

Tested with assembly language, Python, and simple_kernel.

takluyver commented on jupyter_console/interactiveshell.py:
```diff
@@ -109,12 +109,23 @@ class ZMQTerminalInteractiveShell(TerminalInteractiveShell):
     session_id = Unicode()
 
+    def __init__(self, *args, **kwargs):
+        super(TerminalInteractiveShell, self).__init__(*args, **kwargs)
```
`TerminalInteractiveShell` -> `ZMQTerminalInteractiveShell`. The first argument to super() should be the name of the class using it, not the parent class.

dsblank replied:
Indeed, a typo.

takluyver commented on jupyter_console/interactiveshell.py:
```diff
@@ -220,6 +231,43 @@ def handle_execute_reply(self, msg_id, timeout=None):
+            return False, ""
+        ## Handle response:
+        if msg["parent_header"].get("msg_id", None) != msg_id:
+            warn('The kernel did not respond properly to an is_complete_request.')
```
Can we show some details from the message it got - at least msg_type - to make it easier to debug if this does ever occur?

dsblank replied:
This shows the entire message now. Are you looking at an outdated diff?

takluyver commented on jupyter_console/interactiveshell.py:
```diff
@@ -597,3 +644,23 @@ def init_history(self):
+            msg_id = self.client.is_complete("\n".join(self._source_lines_buffered))
+            return self.handle_is_complete_reply(msg_id, timeout=0.05)
+        else:
+            more = (line != "")
```
In case there's one day some auto-indenting interfering with this, I think we should test `line.strip()`, which will be the empty string if line contains only whitespace.

dsblank replied:
I thought about that, and decided that non-empty whitespace text might be meaningful to a language, and so opted to make it be explicitly an empty line. If a system is dealing with indentation, then it should also back-out of the indentation levels one level at a time (perhaps). (I think too much cleverness here can only make the default handler have issues with the wide variety of languages out there.)

takluyver replied:
I don't know of any serious languages where an all-whitespace line has significance, so I'd be inclined to go for it and wait until someone complains. But on the other hand, there's no auto-indent in this case, and there probably never will be, so I guess it's OK.

takluyver commented:
Thanks @dsblank, just a few minor corrections this time.

dsblank commented:
@takluyver Ok, it's all yours! Thanks for your feedback.

takluyver commented:
OK, I'm happy with the code now. One last thing: have you tested this against IPython to ensure that the .05 second timeout is enough for it to get the `is_complete_reply` back under normal circumstances?

dsblank commented:
Yes, I tested on IPython, simple_kernel (no response), and assembly
language--- all with use_kernel_is_complete True and False.

On Tue, Dec 15, 2015 at 3:59 PM, Thomas Kluyver <notifications@github.com>
wrote:

> OK, I'm happy with the code now. One last thing: have you tested this
> against IPython to ensure that the .05 second timeout is enough for it to
> get the is_complete_reply back under normal circumstances?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/jupyter_console/pull/35#issuecomment-164894867>
> .
>

takluyver commented:
Brilliant. Thanks @dsblank

dsblank commented:
Woot! Thanks! :)

takluyver commented on jupyter_console/interactiveshell.py:
```diff
@@ -574,8 +622,7 @@ def interact(self, display_banner=None):
                 try:
-                    self.input_splitter.push(line)
-                    more = self.input_splitter.push_accepts_more()
+                    more, indent = self.handle_source(line)
```
@dsblank we carefully got `indent` back here, but then we haven't actually wired it up to the prompt. Have you got time to have a look? There's a method `_indent_current_str()` on the parent class that I think we'd need to define on this subclass.

dsblank commented:
That would be great to get the indentation working. But I don't know how to define such a method here. What would it do?

takluyver commented:
I think we just need to store and return the value from the last is_complete reply. I'll have a go.

dsblank commented:
Cool; thanks!

takluyver commented:
See #53, which seems to work for me.

dsblank commented:
Yes, that seems to have worked on the other kernels that I had defined indent properly. Now I can define that in others too!
